File: season2/sh/HeapSort.java
package Sort;

public class HeapSort {
	
	public static void heapsort(int[] arr){
		build_MaxHeap(arr);
		for(int i=arr.length; i >=2; i--){
			int temp = arr[1];
			arr[1]=arr[i];
			arr[i]=temp;
			int[] shorted = new int[arr.length-1];
			System.arraycopy(arr, arr[0], shorted, shorted[0], shorted.length);
		}
	}
	public static void build_MaxHeap(int[] arr){
		int heapsize = arr.length-1;
		for(int i =(heapsize/2); i >=1; i--){
			max_heapify(arr,i);
		}
	}

	private static void max_heapify(int[] arr, int i) {
		int heapsize = arr.length-1;
		int largest = i;
		int leftIdx = i * 2;
		int rightIdx = (i*2) + 1;
		
		if(leftIdx <= heapsize && arr[leftIdx] > arr[i]) {
			largest = leftIdx;
		}
		else 
			largest = i;
		
		if(rightIdx <= heapsize && arr[rightIdx] > arr[i]){
			largest = rightIdx;
		}
		if(largest != i){
			int oldlargest = arr[i];
			arr[i] = arr[largest];
			arr[largest] = oldlargest;
			max_heapify(arr,largest);
		}
	}

}

File: season2/heesu/heapSort.js
var test = ['x', 1,7,3,6,5,2];
var test1 = ['x', 10,7,1,6,5,2];
var test2 = ['x', -4,7,3,-6,5,2];


function swap (A, aIdx, bIdx) {
	var temp = A[aIdx];
	A[aIdx] = A[bIdx];
	A[bIdx] = temp;
}


function maxheapify (A, i, end) {
	if (typeof end == "undefined") end = A.length-1;
	var leftIdx = i*2;
	var rightIdx = i*2 + 1;
	var mostIdx = i;
	if (leftIdx <= end && A[leftIdx] > A[mostIdx]) {
		mostIdx = leftIdx;
	}
	if (rightIdx <= end && A[rightIdx] > A[mostIdx]) {
		mostIdx = rightIdx;
	}
	if (i !== mostIdx) {
		swap(A, mostIdx, i);
		maxheapify(A, mostIdx, end);
	}
}


function buildMaxheap (A) {
	var len = A.length;
	for (var i = Math.floor((len-1)/2); i >= 1; i--) {
		maxheapify(A, i);
	}
}


function heapSort (A) {
	buildMaxheap(A);
	// debugger;
	for (var i = A.length - 1; i >= 1; i--) {
		var max = A[i];
		swap(A, i, 1);
		maxheapify(A, 1, i-1);
	}
}